patroni/ha.py

```diff
@@ -368,20 +387,91 @@ def process_healthy_cluster(self):
         return self.follow('demoting self because i do not have the lock and i was a leader',
                            'no action.  i am a secondary and i am following a leader', False)
 
+    def evaluate_scheduled_restart(self):
+        # restart if we need to
+        restart_data = self.future_restart_scheduled()
+        if restart_data:
+            recent_time = self.state_handler.postmaster_start_time()
+            request_time = restart_data['postmaster_start_time']
+            # check if postmaster start time has changed since the last restart
+            if recent_time and request_time and recent_time != request_time:
+                logger.info("Cancelling scheduled restart: postgres restart has already happened at {0}".
+                            format(recent_time))
+                self.delete_future_restart()
+                return
+
+        if (restart_data and
+           self.should_run_scheduled_action('restart', restart_data['schedule'], self.delete_future_restart)):
+            try:
+                ret, message = self.restart(restart_data)
```

Comment:
Restart MUST be done asynchronously...
So far `Ha.restart` method was only called from API, where every request is already performed from a `Thread`, but here it might break ha_loop because restart could be longer than TTL...